src/games/strategy/engine/framework/ui/NewGameChooser.java

```diff
@@ -238,20 +245,29 @@ public synchronized static NewGameChooserModel getNewGameChooserModel() {
     return s_cachedGameModel;
   }
 
-  // any methods touching s_cachedGameModel should be both static and synchronized
+  /** Starts updating the NewGameChooserModel cache, this can be a lengthy process (seconds) */
   public synchronized static void refreshNewGameChooserModel() {
-    clearNewGameChooserModel();
-    s_cachedGameModel = new NewGameChooserModel();
-  }
+    mapParsingInProgress = true;
+    cacheClearedMessenger = new ClearGameChooserCacheMessenger();
 
-  // any methods touching s_cachedGameModel should be both static and synchronized
-  public synchronized static void clearNewGameChooserModel() {
-    if (s_cachedGameModel != null) {
-      s_cachedGameModel.clear();
-      s_cachedGameModel = null;
+    NewGameChooserModel model = new NewGameChooserModel(cacheClearedMessenger);
+      // Cache might not be populated if another thread has cleared
+      // the cache while we were loading.
+    if (mapParsingInProgress) {
```

Comment:
This if statement is really critical to the change. It allows us to remove the synchronization from the clear method. The clear method will instead set this flag and the clear method can then be unblocked to do other stuff. Meanwhile, if clear has been called while we are still building a NewGameChooserModel, then we'll see the flag has been changed and we'll ignore the value of "model" altogether. This is desired, we want that model object to be GC'd so it doesn't consume memory while we are playing.